// @nutui/nutui-react-taro/dist/esm/avatargroup-f3574510.js
import React__default, { useRef } from 'react';
import classNames from 'classnames';
import { A as AvatarContext } from './AvatarContext-c96afdcc.js';
import { c as cn } from './bem-350c1702.js';
import { u as useConfig } from './configprovider-67d19d3a.js';

const defaultProps = {
    maxContent: '',
    maxCount: '',
    maxBgColor: '#eee',
    maxColor: '#666',
    size: '',
    shape: '',
    span: '-8',
    zIndex: 'left',
};
const AvatarGroup = (props) => {
    useConfig();
    const propAvatarGroup = { ...defaultProps, ...props };
    const { className, style, children } = propAvatarGroup;
    const avatarGroupRef = useRef(null);
    const b = cn('avatar-group');
    const cls = classNames(b(''), className);
    const parentAvatar = {
        propAvatarGroup,
        avatarGroupRef,
    };
    return (React__default.createElement(AvatarContext.Provider, { value: parentAvatar },
        React__default.createElement("div", { className: cls, style: style, ref: avatarGroupRef }, children)));
};
AvatarGroup.defaultProps = defaultProps;
AvatarGroup.displayName = 'NutAvatarGroup';

export { AvatarGroup as A };
