// @nutui/nutui-react-taro/dist/esm/backtop.taro-66394a36.js
import React__default, { useState } from 'react';
import { usePageScroll, pageScrollTo } from '@tarojs/taro';
import { I as Icon } from './icon.taro-296a1bec.js';
import { C as ComponentDefaults } from './typings-b9828dba.js';

const defaultProps = {
    ...ComponentDefaults,
    bottom: 20,
    right: 10,
    elId: 'body',
    distance: 200,
    zIndex: 10,
    isAnimation: true,
    duration: 1000,
};
const BackTop = (props) => {
    const { children, bottom, right, elId, distance, zIndex, isAnimation, className, duration, style, onClick, iconClassPrefix, iconFontClassName, } = {
        ...defaultProps,
        ...props,
    };
    const [backTop, SetBackTop] = useState(false);
    // 监听用户滑动页面事件
    usePageScroll((res) => {
        const { scrollTop } = res;
        scrollTop >= distance ? SetBackTop(true) : SetBackTop(false);
    });
    // 返回顶部点击事件
    const goTop = (e) => {
        onClick && onClick(e);
        pageScrollTo({
            scrollTop: 0,
            duration: isAnimation && duration > 0 ? duration : 0,
        });
    };
    const backTopClass = {
        right: `${right}px`,
        bottom: `${bottom}px`,
        zIndex,
    };
    return (React__default.createElement(React__default.Fragment, null,
        React__default.createElement("div", { className: `nut-backtop ${backTop ? 'show' : ''} ${className || ''}`, style: { ...backTopClass, ...style }, onClick: (e) => {
                goTop(e);
            } }, children || (React__default.createElement(Icon, { classPrefix: iconClassPrefix, fontClassName: iconFontClassName, size: "19px", className: "nut-backtop-main", name: "top" })))));
};
BackTop.defaultProps = defaultProps;
BackTop.displayName = 'NutBackTop';

export { BackTop as B };
