// @nutui/nutui-react-taro/dist/esm/badge.taro-a4e59c11.js
import React__default from 'react';
import classNames from 'classnames';
import { I as Icon } from './icon.taro-296a1bec.js';
import { c as cn } from './bem-350c1702.js';
import { C as ComponentDefaults } from './typings-b9828dba.js';

const b = cn('badge');
const defaultProps = {
    ...ComponentDefaults,
    className: '',
    value: '',
    dot: false,
    max: 10000,
    top: '0',
    right: '0',
    zIndex: '0',
    color: '',
    icons: '',
};
const Badge = (props) => {
    const { className, style, children, dot, top, right, zIndex, color, icons, iconClassPrefix, iconFontClassName, } = {
        ...defaultProps,
        ...props,
    };
    function content() {
        if (dot)
            return undefined;
        const { value } = props;
        const { max } = props;
        if (typeof value === 'number' && typeof max === 'number') {
            return max < value ? `${max}+` : value;
        }
        return value;
    }
    const getStyle = () => {
        const style = {};
        style.top = `${Number(top) || parseFloat(top) || 0}px`;
        style.right = `${Number(right) || parseFloat(right) || 0}px`;
        style.zIndex = zIndex;
        style.background = color;
        return style;
    };
    return (React__default.createElement("div", { className: classNames(b(), className), style: style },
        icons !== '' && (React__default.createElement("div", { className: "slot-icons" },
            React__default.createElement(Icon, { classPrefix: iconClassPrefix, fontClassName: iconFontClassName, className: "_icon", name: icons, color: "#ffffff", size: "12" }))),
        React__default.createElement("div", null, children),
        React__default.createElement("div", { className: classNames({ 'is-dot': dot }, b('content'), 'sup'), style: getStyle() }, content())));
};
Badge.defaultProps = defaultProps;
Badge.displayName = 'NutBadge';

export { Badge as B };
