// @nutui/nutui-react-taro/dist/esm/empty.taro-04ae4389.js
import React__default, { useState, useEffect } from 'react';
import { c as cn } from './bem-350c1702.js';
import { u as useConfig } from './configprovider.taro-8d71eb49.js';

/**
 * 内置图片地址
 */
const defaultStatus = {
    empty: 'https://static-ftcms.jd.com/p/files/61a9e3183985005b3958672b.png',
    error: 'https://ftcms.jd.com/p/files/61a9e33ee7dcdbcc0ce62736.png',
    network: 'https://static-ftcms.jd.com/p/files/61a9e31de7dcdbcc0ce62734.png',
};
const defaultProps = {
    description: '无内容',
    image: 'empty',
    imageSize: '',
    className: '',
};
const Empty = (props) => {
    const { locale } = useConfig();
    defaultProps.description = locale.noData || defaultProps.description;
    const { image, imageSize, description, children, className, ...rest } = {
        ...defaultProps,
        ...props,
    };
    const [imgStyle, setImgStyle] = useState({});
    const b = cn('empty');
    let imageNode = image;
    if (typeof image === 'string') {
        const isHttpUrl = image.startsWith('https://') ||
            image.startsWith('http://') ||
            image.startsWith('//');
        const imageUrl = isHttpUrl ? image : defaultStatus[image];
        imageNode = React__default.createElement("img", { className: "img", src: imageUrl, alt: "empty" });
    }
    useEffect(() => {
        setImgStyle(() => {
            if (!imageSize) {
                return {};
            }
            if (typeof imageSize === 'number') {
                return {
                    width: `${imageSize}px`,
                    height: `${imageSize}px`,
                };
            }
            return {
                width: imageSize,
                height: imageSize,
            };
        });
    }, [imageSize]);
    return (React__default.createElement("div", { className: `${b()} ${className}`, ...rest },
        React__default.createElement(React__default.Fragment, null,
            React__default.createElement("div", { className: b('image'), style: imgStyle }, imageNode),
            typeof description === 'string' ? (React__default.createElement("div", { className: b('description') }, description)) : ({ description }),
            children)));
};
Empty.defaultProps = defaultProps;
Empty.displayName = 'NutEmpty';

export { Empty as E };
