// @nutui/nutui-react-taro/dist/esm/fixednav.taro-1f05acaf.js
import React__default from 'react';
import classNames from 'classnames';
import { I as Icon } from './icon.taro-296a1bec.js';
import { O as Overlay } from './overlay.taro-21e87fb1.js';
import { c as cn } from './bem-350c1702.js';
import { u as useConfig } from './configprovider.taro-8d71eb49.js';
import { C as ComponentDefaults } from './typings-b9828dba.js';

const defaultProps = {
    ...ComponentDefaults,
    fixednavClass: 'nut-fixednav',
    activeText: '',
    unActiveText: '',
    type: 'right',
    position: {
        top: 'auto',
        bottom: 'auto',
    },
};
const FixedNav = (props) => {
    const { locale } = useConfig();
    const { fixednavClass, overlay, visible, navList, activeText, unActiveText, position, onChange, onSelected, type, slotList, slotBtn, iconClassPrefix, iconFontClassName, ...rest } = {
        ...defaultProps,
        ...props,
    };
    const b = cn('fixednav');
    const classes = classNames({
        active: visible,
    }, type, fixednavClass, b(''));
    const onSelectCb = (event, item) => {
        onSelected(item, event);
    };
    const onUpdateValue = (value = !visible) => {
        onChange(value);
    };
    return (React__default.createElement("div", { className: classes, style: position, ...rest },
        overlay && (React__default.createElement(Overlay, { visible: visible, zIndex: 200, onClick: () => onUpdateValue(false) })),
        React__default.createElement("div", { className: "list" }, slotList || (React__default.createElement("div", { className: "nut-fixednav__list" }, navList.map((item, index) => {
            return (React__default.createElement("div", { className: "nut-fixednav__list-item", onClick: (event) => onSelectCb(event, item), key: item.id || index },
                React__default.createElement("img", { src: item.icon, alt: "" }),
                React__default.createElement("div", { className: "nut-fixednav__list-text" }, item.text),
                item.num && React__default.createElement("div", { className: "b" }, item.num)));
        })))),
        React__default.createElement("div", { className: "nut-fixednav__btn", onClick: () => onUpdateValue() }, slotBtn || (React__default.createElement(React__default.Fragment, null,
            React__default.createElement(Icon, { classPrefix: iconClassPrefix, fontClassName: iconFontClassName, name: "left", color: "#fff" }),
            React__default.createElement("div", { className: "text" }, visible
                ? activeText || locale.fixednav.activeText
                : unActiveText || locale.fixednav.unActiveText))))));
};
FixedNav.defaultProps = defaultProps;
FixedNav.displayName = 'NutFixedNav';

export { FixedNav as F };
