// @nutui/nutui-react-taro/dist/esm/indicator.taro-cba8b010.js
import React__default, { useEffect } from 'react';
import classNames from 'classnames';
import { c as cn } from './bem-350c1702.js';

const defaultProps = {
    size: 3,
    current: 1,
    block: false,
    align: 'center',
    fillZero: true,
    vertical: false,
};
const Indicator = (props) => {
    const { size, current, block, align, fillZero, children, className, vertical, ...rest } = {
        ...defaultProps,
        ...props,
    };
    const b = cn('indicator');
    const classes = classNames({
        [`${b('block')}`]: block,
        [`${b('align')}__${align}`]: block && align,
        [`${b('vertical')}`]: vertical,
    }, b(''));
    const renderEles = () => {
        const childs = [];
        for (let item = 1; item <= size; item++) {
            childs.push(item === current ? (React__default.createElement("div", { key: item, className: b('number') }, fillZero && item < 10 ? `0${item}` : item)) : (React__default.createElement("div", { key: item, className: b('dot') })));
        }
        return childs;
    };
    useEffect(() => { }, []);
    return (React__default.createElement("div", { className: classNames(classes, className), ...rest }, renderEles()));
};
Indicator.defaultProps = defaultProps;
Indicator.displayName = 'NutIndicator';

export { Indicator as I };
