// @nutui/nutui-react-taro/dist/esm/row.taro-f64d4a13.js
import React__default from 'react';
import classNames from 'classnames';
import { D as DataContext } from './UserContext-0aac9a82.js';

const defaultProps = {
    type: '',
    justify: 'start',
    align: 'flex-start',
    wrap: 'nowrap',
    gutter: '0',
};
const Row = (props) => {
    const { className, style = {}, children, type, justify, align, wrap, gutter, onClick, } = {
        ...defaultProps,
        ...props,
    };
    const prefixCls = 'nut-row';
    const getClass = (prefix, type) => {
        const classType = type ? `nut-row-${prefix}-${type}` : '';
        const className = prefix ? classType : `nut-row-${type}`;
        return className;
    };
    const getClasses = () => {
        return `
    ${getClass('', type)}
    ${getClass('justify', justify)}
    ${getClass('align', align)}
    ${getClass('flex', wrap)}
    ${prefixCls}
   `;
    };
    const parentRow = {
        gutter,
    };
    return (React__default.createElement(DataContext.Provider, { value: parentRow }, React__default.createElement('div', {
        className: classNames(getClasses(), className),
        style,
        onClick: (e) => {
            onClick && onClick(e, 'row');
        },
    }, children)));
};
Row.defaultProps = defaultProps;
Row.displayName = 'NutRow';

export { Row as R };
