// @nutui/nutui-react-taro/dist/esm/skeleton.taro-76faf63f.js
import React__default from 'react';
import classNames from 'classnames';
import { A as Avatar } from './avatar.taro-3676f38f.js';
import { c as cn } from './bem-350c1702.js';

const defaultProps = {
    width: '100px',
    height: '100px',
    row: 1,
    animated: false,
    title: false,
    avatar: false,
    round: false,
    avatarSize: '50px',
    loading: false,
    avatarShape: 'round',
};
const Skeleton = (props) => {
    const { className, width, height, animated, row, title, avatar, avatarSize, round, loading, children, avatarShape, ...restProps } = {
        ...defaultProps,
        ...props,
    };
    const b = cn('skeleton');
    const classes = classNames(className, b());
    const blockClass = classNames({
        blockClass: true,
        'blockClass--round': round,
    });
    const avatarClass = classNames({
        avatarClass: true,
        [`avatarClass--${avatarShape}`]: avatarShape,
    });
    const repeatLines = (num) => {
        return Array.from({ length: num }, (v, i) => i);
    };
    const getStyle = () => {
        if (avatarSize) {
            return {
                width: avatarSize,
                height: avatarSize,
            };
        }
        return {
            width: '50px',
            height: '50px',
        };
    };
    return (React__default.createElement(React__default.Fragment, null, loading ? (React__default.createElement("div", null, children)) : (React__default.createElement("div", { className: classes, ...restProps },
        animated && React__default.createElement("div", { className: "skeleton-animation" }),
        React__default.createElement("div", { className: "nut-skeleton-content" },
            avatar && (React__default.createElement(Avatar, { className: avatarClass, bgColor: "rgb(239, 239, 239)", shape: avatarShape, style: getStyle() })),
            row === 1 && (React__default.createElement("div", { className: blockClass, style: { width, height } })),
            React__default.createElement("div", { className: "skeleton-content-line" },
                title && React__default.createElement("div", { className: "skeleton-title" }),
                repeatLines(row).map((item, index) => {
                    return (React__default.createElement("div", { className: blockClass, key: index, style: { width, height } }));
                })))))));
};
Skeleton.defaultProps = defaultProps;
Skeleton.displayName = 'NutSkeleton';

export { Skeleton as S };
