// @nutui/nutui-react-taro/dist/esm/switch.taro-34ca8228.js
import React__default, { useState, useEffect } from 'react';
import { c as cn } from './bem-350c1702.js';

const defaultProps = {
    isAsync: false,
    checked: false,
    disable: false,
    activeColor: '',
    inactiveColor: '',
    activeText: '',
    inactiveText: '',
    className: '',
};
const Switch = (props) => {
    const { isAsync, checked, disable, activeColor, inactiveColor, activeText, inactiveText, onChange, className, style, } = {
        ...defaultProps,
        ...props,
    };
    const [value, setValue] = useState(false);
    const b = cn('switch');
    useEffect(() => {
        setValue(checked);
    }, [checked]);
    const classes = () => {
        return `${b()} ${value ? 'switch-open' : 'switch-close'} ${disable ? `${b()}-disable` : ''} ${`${b()}-base`} ${className}`;
    };
    const styles = () => {
        const myStyle = {
            backgroundColor: value ? activeColor : inactiveColor,
            ...(style || {}),
        };
        return myStyle;
    };
    const onClick = (event) => {
        if (disable)
            return;
        if (!isAsync) {
            setValue(!value);
        }
        onChange && onChange(!value, event);
    };
    return (React__default.createElement("div", { className: classes(), onClick: (e) => onClick(e), style: styles() },
        React__default.createElement("div", { className: "switch-button" },
            !value && React__default.createElement("div", { className: "close-line" }),
            activeText && (React__default.createElement(React__default.Fragment, null, value ? (React__default.createElement("div", { className: `${b('label')} open` }, activeText)) : (React__default.createElement("div", { className: `${b('label')} close` }, inactiveText)))))));
};
Switch.defaultProps = defaultProps;
Switch.displayName = 'NutSwitch';

export { Switch as S };
