// @nutui/nutui-react-taro/dist/esm/timedetail.taro-5b2f2269.js
import React__default, { useState, useEffect } from 'react';
import { c as cn } from './bem-350c1702.js';
import { u as useConfig } from './configprovider.taro-8d71eb49.js';

const defaultProps = {
    className: '',
    currentKey: 0,
    currentTime: [],
    times: [],
    select: () => null,
};
const TimeDetail = (props) => {
    useConfig();
    const { children, times, className, currentKey, currentTime, select } = {
        ...defaultProps,
        ...props,
    };
    const b = cn('timedetail');
    const [renderData, setRenderData] = useState([]);
    useEffect(() => {
        const currentData = times.find((timesItem) => String(timesItem.key) === String(currentKey));
        const renderData = currentData ? currentData.list : [];
        // 根据选中的日期回显当前日期可配送的时间
        setRenderData(renderData);
    }, [times, currentKey]);
    // 选中时间的回调
    const handleTime = (time) => {
        select(time);
    };
    // 选中的配送时间增加 active 类名
    const getDetailClass = (item) => {
        let initClass = 'nut-timedetail__item';
        const curTimeData = currentTime.find((item) => String(item.key) === String(currentKey));
        if (curTimeData && curTimeData.list && curTimeData.list.includes(item)) {
            initClass += ' nut-timedetail__item-active';
        }
        return initClass;
    };
    return (React__default.createElement("div", { className: `${b()} ${className || ''}` }, renderData.map((item, index) => (React__default.createElement("span", { className: getDetailClass(item), key: item, onClick: () => handleTime(item) }, item)))));
};
TimeDetail.defaultProps = defaultProps;
TimeDetail.displayName = 'NutTimeDetail';

export { TimeDetail as T };
