// @nutui/nutui-react-taro/dist/esm/toast.taro-d6c14811.js
import React__default, { useState, useEffect } from 'react';
import classNames from 'classnames';
import { c as cn } from './bem-350c1702.js';
import { I as Icon } from './icon-296a1bec.js';

const defaultProps = {
    id: '',
    style: {},
    icon: null,
    iconSize: '20',
    msg: '',
    bottom: '30px',
    duration: 2,
    center: true,
    type: 'text',
    title: '',
    customClass: '',
    size: 'base',
    textAlignCenter: true,
    bgColor: 'rgba(0, 0, 0, .8)',
    cover: false,
    coverColor: 'rgba(0, 0, 0, 0)',
    closeOnClickOverlay: false,
    visible: false,
    onClose: () => { }, // 未实现
};
// export default class Notification extends React.PureComponent<NotificationProps> {
const Toast = (props) => {
    const { children, id, style, icon, iconSize, msg, bottom, duration, center, type, title, customClass, size, textAlignCenter, bgColor, cover, coverColor, closeOnClickOverlay, visible, onClose, ...rest } = { ...defaultProps, ...props };
    let timer;
    const [openState, SetOpenState] = useState(false);
    useEffect(() => {
        if (visible) {
            SetOpenState(true);
            show();
        }
        else {
            hide();
        }
    }, [visible]);
    const clearTimer = () => {
        if (timer) {
            clearTimeout(timer);
            timer = null;
        }
    };
    const hide = () => {
        SetOpenState(false);
        onClose();
    };
    const show = () => {
        clearTimer();
        if (duration) {
            timer = window.setTimeout(() => {
                hide();
            }, duration * 1000);
        }
    };
    const clickCover = () => {
        if (closeOnClickOverlay) {
            hide();
        }
    };
    const toastBem = cn('toast');
    const hasIcon = () => {
        if (type !== 'text') {
            return true;
        }
        return !!icon;
    };
    const iconName = () => {
        if (icon) {
            return icon;
        }
        return {
            success: 'success',
            fail: 'failure',
            warn: 'tips',
            loading: 'loading',
        }[type];
    };
    const classes = classNames({
        'nut-toast-center': center,
        'nut-toast-has-icon': icon,
        'nut-toast-cover': cover,
        'nut-toast-loading': type === 'loading',
        [`${customClass}`]: true,
        [`nut-toast-${size}`]: true,
    });
    return (React__default.createElement(React__default.Fragment, null, openState ? (React__default.createElement("div", { className: `${toastBem()} ${classes}`, id: id, style: {
            bottom: center ? 'auto' : `${bottom}`,
            backgroundColor: cover ? coverColor : '',
            ...style,
        }, onClick: () => {
            clickCover();
        } },
        React__default.createElement("div", { className: toastBem('inner'), style: {
                textAlign: textAlignCenter ? 'center' : 'left',
                backgroundColor: bgColor,
            } },
            hasIcon() ? (React__default.createElement("p", { className: toastBem('icon-wrapper') },
                React__default.createElement(Icon, { name: iconName(), color: "#ffffff", size: iconSize }))) : null,
            title ? React__default.createElement("div", { className: "nut-toast-title" }, title) : null,
            React__default.createElement("span", { className: toastBem('text') }, msg)))) : null));
};
Toast.defaultProps = defaultProps;
Toast.displayName = 'NutToast';

export { Toast as T };
